storage/src/adb/any/variable/sync.rs

```diff
@@ -86,140 +87,52 @@ pub(crate) async fn init_journal<E: Storage + Metrics, V: Codec>(
     }
 
-    // Remove any items beyond upper_bound
-    truncate_upper_section(&mut journal, upper_bound, items_per_section).await?;
-
-    Ok(journal)
-}
-
-/// Remove items beyond the `upper_bound` location (inclusive).
-/// Assumes each section contains `items_per_section` items.
-async fn truncate_upper_section<E: Storage + Metrics, V: Codec>(
-    journal: &mut VJournal<E, V>,
-    upper_bound: u64,
-    items_per_section: u64,
-) -> Result<(), Error> {
-    // Find which section contains the upper_bound item
-    let upper_section = upper_bound / items_per_section;
-    let Some(blob) = journal.blobs.get(&upper_section) else {
-        return Ok(()); // Section doesn't exist, nothing to truncate
-    };
-
-    // Calculate the logical item range for this section
-    let section_start = upper_section * items_per_section;
-    let section_end = section_start + items_per_section - 1;
-
-    // If upper_bound is at the very end of the section, no truncation needed
-    if upper_bound >= section_end {
-        return Ok(());
+    let size = get_size(&journal, items_per_section).await?;
+    if size > upper_bound + 1 {
+        return Err(adb::Error::UnexpectedData(size));
     }
 
-    // Calculate how many items to keep (upper_bound is inclusive)
-    let items_to_keep = (upper_bound - section_start + 1) as u32;
-    debug!(
-        upper_section,
-        upper_bound,
-        section_start,
-        section_end,
-        items_to_keep,
-        "truncating section to remove items beyond upper_bound"
-    );
-
-    // Find where to rewind to (after the last item we want to keep)
-    let target_byte_size = compute_offset::<E, V>(
-        blob,
-        &journal.cfg.codec_config,
-        journal.cfg.compression.is_some(),
-        items_to_keep,
-    )
-    .await?;
-
-    // Rewind to the appropriate position to remove items beyond the upper bound
-    journal
-        .rewind_section(upper_section, target_byte_size)
-        .await?;
-
-    debug!(
-        upper_section,
-        items_to_keep, target_byte_size, "section truncated"
-    );
-
-    Ok(())
+    Ok((journal, size))
 }
 
-/// Return the byte offset of the next element after `items_count` elements of `blob`.
-async fn compute_offset<E: Storage + Metrics, V: Codec>(
-    blob: &commonware_runtime::buffer::Append<E::Blob>,
-    codec_config: &V::Cfg,
-    compressed: bool,
-    items_count: u32,
-) -> Result<u64, Error> {
-    use crate::journal::variable::{Journal, ITEM_ALIGNMENT};
-
-    if items_count == 0 {
+/// Returns the number of items in the journal.
+pub(crate) async fn get_size<E: Storage + Metrics, V: Codec>(
+    journal: &VJournal<E, V>,
+    items_per_section: u64,
+) -> Result<u64, adb::Error> {
+    let Some(last_section) = journal.blobs.last_key_value().map(|(&s, _)| s) else {
         return Ok(0);
+    };
+    let last_section_start = last_section * items_per_section;
+    let stream = journal.replay(last_section, 0, NZUsize!(1024)).await?;
```

Comment:
nit: prefer this replay param to be a const like other `adb`s: https://github.com/commonwarexyz/monorepo/blob/d43dc6e7525464403d77e36030b29aa49c337ec6/storage/src/adb/keyless.rs#L33-L34